gaps/gap-24_agreement_permissions_management/gap-24_agreement_permissions_management.md

```diff
@@ -0,0 +1,326 @@
+
+**Benefits:**
+- Simple implementation - no automatic info propagation is required from the network
+- Deterministic - Agent App action is required to get all relevant information to the Grantee node, and the operation can be synchronous. This leaves little room for "undefined behaviour".
+
+**Drawbacks:**
+- Requires the Requestor Agent App to obtain the `AgreementId` via other means, eg. via some sort of a "supervisor protocol"
+- Requires implementation of dedicated "synchronization" API, which is quite implementation-specific, and makes the Golem's low-level APIs less abstract.
+
+2. "Auto-propagated" Agreement synchronization
+
+This implementation ensures that Golem Network protocols ensure the `AgreementId->ProviderId` info is automatically propagated to relevant nodes in the network. 
+
+Possible implementation requires the following:
+  1. Each `setAgreementPermissions` processed by the Provider node triggers `AgreementId->ProviderId` messages to all entitled Requestors.
+    - note this is not sufficient if a Grantee node is still offline at that point, therefore the following active notifications are required.
+  2. Each entitled mode which has an ACL maintains a list of entitled Grantees.
+  3. All golem nodes use "GAP-30 Node presence notification" concept to broadcast its presence to whole network (this must happen each time the node joins or re-joins the network, eg. after connectivity failure)
+  4. Whenever a node receives a "Node resence notification" it checks its entitled Grantees list - if the sender of "presence notification" is on the Grantee list - reply with `AgreementId->ProviderId`.
+
+
+**Benefits:**
+- Developer UX. The ACL propagation is "transparent" to nodes involved in Agreement maintenance - ie. any node which becomes entitled to manage an Agreement and its related entities is automatically able to call respective APIs (eg. `GetAgreement` passing the `AgreementId`) and is expected to receive relevant details.
+- No extension of low-level APIs is required - the APIs remain generic and abstract.
+
+**Drawbacks:**
+- Implementation complexity.
+- Quasi-deterministic. As the implementation based on "GAP-30 Node presence notification" mechanism includes message propagation delay, and does not guarantee reliable message delivery - the possibility of relevant nodes not receiving the `AgreementId->ProviderId` mapping cannot be ignored. There may be a period of "unawareness" when the Agent App on Grantee node already "knows" it can manage a specific Agreement, but it is unable to call `getAgreement` because the respective `yagna` daemon hasn't yet received the owning Provider's identifier from the network.
+- Only possible when Grantee Id can be inferred from the ACL. Some attestation types (eg. based on X.509 certificates) do not explicitly indicate entitled node Ids. Therefore it may not always be possible to determine Grantee nodes and notify them appropriately.
+
+**Recommendation:** Implementation 1. 
+
```

Comment:
Add the following aspect:

3. Event propagation

Provider is expected to actively notify a set of Grantees with Agreement (and respective entity) events. However the list of recipient nodes for events needs to be maintained by the Provider. Following logic is proposed:
- A successful call to `attachAgreement` from a granteeId will add this Grantee to the list of event "subscribers"
  - A new grantee added to "subscribers" list triggers sending full history of relevant events to that grantee (* do we want to optimize this, eg. for a grantee who was offline and re-connects itself?)
- Each event shall be forwarded by the Provider to all recipients from "event subscriber" list
  - In case event message delivery is failing for a given grantee (delivery confirmation to be decided during implementation) - that grantee is removed from "event subscriber" list
- Whenever Grantee disconnects, and then reconnects itself - it must call the `attachAgreement` again
- In situations where Grantee isn't aware it has been removed from "subscriber list" (eg. network went down, event message delivery failed, Provider removed the Grantee from list)
  - (Option 1) A simple "health check" mechanism is proposed - where Grantee side, for all Agreements in non-terminated state, performs a periodic "am I still subscribed" call to Provider. If False is returned, `yagna` daemon calls `attachAgreement` again, in order to get the latest state of the Agreement. (This option is specific to Agreement synchronization and is not a generic mechanism)
  - (Option 2) Implement message delivery guarantee on GSB transport level, which would need to have following features:
    - Message delivery can be either guaranteed or not guaranteed (as required)
    - Message delivery sequence can be either guaranteed or not guaranteed (as required)
      - User code is able to specify the message sequence indicator (as per functionality, eg. specify that Agreement timestamp is to be used as sequence indicator)
    If the above is available, the event message delivery is guaranteed, and the upper layer (Agreement-specific) does not need to worry about this.
**Option 2 is definitely a more substantial change, and requires a broader discussion... it probably should be considered a separate GAP...**